adapters/tlsio_ssl_esp8266.c

```diff
@@ -0,0 +1,712 @@
+    return sock_errno;
+}
+
+static void lwip_set_non_block(int fd) 
+{
+  int flags = -1;
+  int error = 0;
+
+  while(1){
+      flags = fcntl(fd, F_GETFL, 0);
+      if (flags == -1){
+          error = lwip_net_errno(fd);
+          if (error != EINTR){
+              break;
+          }
+      } else{
+          break;
+      }
+  }
+
+  while(1){
+      flags = fcntl(fd, F_SETFL, flags | O_NONBLOCK);
+      if (flags == -1) {
+          error = lwip_net_errno(fd);
+          if (error != EINTR){
+              break;
+          }
+      } else{
+          break;
+      }
+  }
+
+}
+
+
+LOCAL void openssl_thread_LWIP_CONNECTION(void *p)
+{
+    int ret = -1;
+
+    int sock = -1;
+    struct sockaddr_in sock_addr;
+    fd_set readset;
+    fd_set writeset;
+    fd_set errset;
+
+    ip_addr_t target_ip;
+    SSL_CTX *ctx;
+    SSL *ssl;
+
+    TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)p;
+
+    LogInfo("OpenSSL thread start...\n");
+
+    do {
+        ret = netconn_gethostbyname(tls_io_instance->hostname, &target_ip);
+    } while(ret);
+    // This is useful for debugging purpose.
+    // LogInfo("get target IP is %d.%d.%d.%d\n", (unsigned char)((target_ip.addr & 0x000000ff) >> 0),
+    //                                             (unsigned char)((target_ip.addr & 0x0000ff00) >> 8),
+    //                                             (unsigned char)((target_ip.addr & 0x00ff0000) >> 16),
+    //                                             (unsigned char)((target_ip.addr & 0xff000000) >> 24));
+
+    LogInfo("create socket ......");
+    sock = socket(AF_INET, SOCK_STREAM, 0);
+    if (sock < 0) {
+        LogError("create socket failed\n");
+        return;
+    }
+    LogInfo("create socket OK\n");
+
+    LogInfo("bind socket ......");
+    memset(&sock_addr, 0, sizeof(sock_addr));
+    sock_addr.sin_family = AF_INET;
+    sock_addr.sin_addr.s_addr = 0;
+    sock_addr.sin_port = htons(OPENSSL_LOCAL_TCP_PORT);
+    ret = bind(sock, (struct sockaddr*)&sock_addr, sizeof(sock_addr));
+    if (ret) {
+        LogError("bind socket failed\n");
+        return;
+    }
+    LogInfo("bind socket OK\n");
+
+    LogInfo("socket connect to remote ......");
+    memset(&sock_addr, 0, sizeof(sock_addr));
+    sock_addr.sin_family = AF_INET;
+    sock_addr.sin_addr.s_addr = target_ip.addr;
+    sock_addr.sin_port = htons(tls_io_instance->port);
+
+    lwip_set_non_block(sock);
+
+    ret = connect(sock, (struct sockaddr*)&sock_addr, sizeof(sock_addr));
+    if (ret == -1) {
+        ret = lwip_net_errno(sock);
+        if (ret != EINPROGRESS){
+            LogError("socket connect failed %s\n", tls_io_instance->hostname);
+            return;
+        } else{
+            LogInfo("socket connect in progress\n");
+        }
+    }
+    LogInfo("socket connect OK\n");
+
+    char recv_buf[128];
+    int recv_bytes = (int)sizeof(recv_buf);
+    int retry = 0;
+
+    while (retry < 10){
```

Comment:
Please consider create a constant for the number of retries besides use directly the number 10.